jetty-ee9/jetty-ee9-annotations/src/main/java/org/eclipse/jetty/ee9/annotations/AnnotationParser.java

```diff
@@ -578,31 +577,38 @@ public void parse(final Set<? extends Handler> handlers, Resource r) throws Exce
      */
     protected void parseDir(Set<? extends Handler> handlers, Resource dirResource) throws Exception
     {
-        Path dir = dirResource.getPath();
-
         if (LOG.isDebugEnabled())
-            LOG.debug("Scanning dir {}", dir.toUri());
+            LOG.debug("Scanning dir {}", dirResource);
+
+        assert dirResource.isDirectory();
 
         ExceptionUtil.MultiException multiException = new ExceptionUtil.MultiException();
-        try (Stream<Path> classStream = Files.walk(dir))
-        {
-            classStream
-                .filter(Files::isRegularFile)
-                .filter((path) -> !FileID.isHidden(dir, path))
-                .filter(FileID::isNotMetaInfVersions)
-                .filter(FileID::isNotModuleInfoClass)
-                .filter(FileID::isClassFile)
-                .forEach(classFile ->
-                {
-                    try
-                    {
-                        parseClass(handlers, dirResource, classFile);
-                    }
-                    catch (Exception ex)
-                    {
-                        multiException.add(new RuntimeException("Error scanning entry " + ex, ex));
-                    }
-                });
+
+        for (Resource candidate : dirResource.getAllResources())
```

Comment:
Is this testable?

Reply:
Yes, there are tests for getAllResources and Jan added an integration test that checked the bug that was in this loop that started this whole process off.